src/EventuallyQueue.js

```diff
@@ -0,0 +1,373 @@
+  queueId: string,
+  action: string,
+  object: ParseObject,
+  serverOptions: SaveOptions | RequestOptions,
+  id: string,
+  className: string,
+  hash: string,
+  createdAt: Date,
+};
+
+type Queue = Array<QueueObject>;
+
+const QUEUE_KEY = 'Parse/Eventually/Queue';
+let queueCache = [];
+let dirtyCache = true;
+let polling = undefined;
+
+/**
+ * Provides utility functions to queue objects that will be
+ * saved to the server at a later date.
+ *
+ * @class Parse.EventuallyQueue
+ * @static
+ */
+const EventuallyQueue = {
+  /**
+   * Add object to queue with save operation.
+   *
+   * @function save
+   * @name Parse.EventuallyQueue.save
+   * @param {ParseObject} object Parse.Object to be saved eventually
+   * @param {object} [serverOptions] See {@link https://parseplatform.org/Parse-SDK-JS/api/master/Parse.Object.html#save Parse.Object.save} options.
+   * @returns {Promise} A promise that is fulfilled if object is added to queue.
+   * @static
+   * @see Parse.Object#saveEventually
+   */
+  save(object: ParseObject, serverOptions: SaveOptions = {}): Promise {
+    return this.enqueue('save', object, serverOptions);
+  },
+
+  /**
+   * Add object to queue with save operation.
+   *
+   * @function destroy
+   * @name Parse.EventuallyQueue.destroy
+   * @param {ParseObject} object Parse.Object to be destroyed eventually
+   * @param {object} [serverOptions] See {@link https://parseplatform.org/Parse-SDK-JS/api/master/Parse.Object.html#destroy Parse.Object.destroy} options
+   * @returns {Promise} A promise that is fulfilled if object is added to queue.
+   * @static
+   * @see Parse.Object#destroyEventually
+   */
+  destroy(object: ParseObject, serverOptions: RequestOptions = {}): Promise {
+    return this.enqueue('destroy', object, serverOptions);
+  },
+
+  /**
+   * Generate unique identifier to avoid duplicates and maintain previous state.
+   *
+   * @param {string} action save / destroy
+   * @param {object} object Parse.Object to be queued
+   * @returns {string}
+   * @static
+   * @ignore
+   */
+  generateQueueId(action: string, object: ParseObject): string {
+    object._getId();
+    const { className, id, _localId } = object;
+    const uniqueId = object.get('hash') || _localId;
+    return [action, className, id, uniqueId].join('_');
+  },
+
+  /**
+   * Build queue object and add to queue.
+   *
+   * @param {string} action save / destroy
+   * @param {object} object Parse.Object to be queued
+   * @param {object} [serverOptions]
+   * @returns {Promise} A promise that is fulfilled if object is added to queue.
+   * @static
+   * @ignore
+   */
+  async enqueue(
+    action: string,
+    object: ParseObject,
+    serverOptions: SaveOptions | RequestOptions
+  ): Promise {
+    const queueData = await this.getQueue();
+    const queueId = this.generateQueueId(action, object);
+
+    let index = this.queueItemExists(queueData, queueId);
+    if (index > -1) {
+      // Add cached values to new object if they don't exist
```

Comment:
If you try to add an object to the queue that already exists what should happen? Should it override the existing object?

Reply:
Not sure about overriding. Couldn't that contradict the queue's FIFO? Unless the two operations to be merged are also adjacent to each other, then that operation could be optimized. However, this would be a custom flavor because a separate create and update operation are not necessarily the same as a merged create op.

Reply:
It sends to the server as a queue. It currently merges for optimization. I’m not sure if the way I’m merging is correct. I do have a test for this.

Reply:
https://github.com/parse-community/Parse-SDK-JS/pull/1291/files#diff-75a47fd2940c491cf43ac004fdfa07d1324e16064dab5a241132941fa949ef5cR141

What should the output be?

Reply:
Sorry for this criss-cross comment... https://github.com/parse-community/Parse-SDK-JS/pull/1291#discussion_r571507883